### DISCUSSION TITLE
Use Skywalking with Spring Native

### DISCUSSION THREAD

**Discussion by mattwmj** [apache/skywalking#7640] [Ideas]
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [1 ] Question or discussion
- [ ] Bug
- [ ] Requirement
- [ ] Feature or performance improvement

___
### Question
- What do you want to know?

We are using Skywalking java agent with Spring Boot and they work together very well.

Recently, I am looking into Spring Native to speed up Spring Boot in Docker container.

https://docs.spring.io/spring-native/docs/current/reference/htmlsingle/

However, It seems that Spring Native and GraalVM do not support java agent currently.

https://github.com/oracle/graal/issues/1065

I understand this defect is not about Skywalking, but is there a solution to use Skywalking when java agent is not an option?

Is there a feature that I can instrument java class with Skywalking during compile time?

Thanks!

**Comment by wu-sheng** [comment]
In theory, of course, you could. We keep our eyes on all static compiling solution for a while. But the community don't have anyone saying he/she wants to lead on that part.

**Comment by mattwmj** [comment]
> In theory, of course, you could. We keep our eyes on all static compiling solution for a while. But the community don't have anyone saying he/she wants to lead on that part.

Is there any document about Skywalking static compiling that I can start with?

**Comment by wu-sheng** [comment]
There isn't. No PMC member is on this. I think you could learn how Java compiling plugin works, such as maven system.
BTW, this solution also could work for Android. So, it is valuable, but complex.

**Comment by linghengqian** [comment]
@mattwmj 
- I don't intend to open a new discussion, so I'm hanging some pending things under this discussion to identify you with some PRs that might change. 

- Spring Native has entered maintenance, so I'll go straight to `org.springframework.boot:spring-boot-maven-plugin:3.0.0-M4` instead of talking about `org.springframework.boot:spring-boot-maven-plugin:2.x` . 

- `spring-boot-maven-plugin:3.0.0-M4` is packaged  GraalVM Native Image with Paketo BuildPacks, when you configure `-javaagent:app/META-INF/skywalking-agent.jar` via environment variable `JAVA_OPTS`, **what error log do you see**? The configuration in Paketo BuildPacks should be similar to https://github.com/cloudfoundry/java-buildpack/issues/555#issuecomment-368583041 .
